### PR TITLE
Enhancements to `setupTLSConfig`

### PR DESCRIPTION
**Pull request by simon-wenmouth**
The function `setupTLSConfig` is now implemented in terms of the
function `SetupTLSConfig` in `consul/api`.

The function also respects two additional configuration properties.
- `tls_server_name` (optional) which sets the server name to use as
  the SNI host when connecting via TLS (for when it differs from the
  server name in `address`). Has the same behavior as the environment
  variable `CONSUL_TLS_SERVER_NAME`.
- `tls_ca_path` (optional) which sets the path to a directory of CA
  certificates to use for Consul communication. Has the same behavior
  as the environment variable `CONSUL_CAPATH`.

The function also now logs:
- a debug message containing the values used to create the `Config`
  from the `crypto/tls` package.
- a warning message when only one of `tls_key_file` or `tls_cert_file`
  are provided in the `conf` map.

### CODE REVIEW

File: physical/consul/consul.go [hunk 1]
```diff
@@ -240,65 +238,80 @@ func NewConsulBackend(conf map[string]string, logger log.Logger) (physical.Backe
...
+func setupTLSConfig(conf map[string]string, tlsConfig *api.TLSConfig, logger log.Logger) (*tls.Config, error) {
+	address, ok := conf["address"]
+	if ok {
+		tlsConfig.Address = address
 	}
 
-	insecureSkipVerify := false
-	if _, ok := conf["tls_skip_verify"]; ok {
-		insecureSkipVerify = true
+	serverName, ok := conf["tls_server_name"]
```
Comment (jefferai) [inline physical/consul/consul.go]: The default behavior still needs to be the net.SplitHostPort behavior, if not overridden with tls_server_name.

File: website/source/docs/configuration/storage/consul.html.md [hunk 2]
```diff
@@ -100,6 +100,13 @@ connection. You can read more about encrypting Consul connections on the
   [`ca_file`](https://www.consul.io/docs/agent/options.html#ca_file) setting in
   Consul.
 
+- `tls_ca_path` `(string: "")` – Specifies the path to a directory of
+  certificate authority files for Consul communication.  Is an alternative to
+  `tls_ca_file`.  This defaults to the system bundle if not specified.
```
Comment (jefferai) [inline website/source/docs/configuration/storage/consul.html.md]: Duplicate tls_ca_file value.

File: physical/consul/consul.go [hunk 3]
```diff
@@ -240,65 +238,80 @@ func NewConsulBackend(conf map[string]string, logger log.Logger) (physical.Backe
...
-		tlsClientConfig.Certificates = []tls.Certificate{tlsCert}
 	}
 
-	if tlsCaFile, ok := conf["tls_ca_file"]; ok {
-		caPool := x509.NewCertPool()
-
-		data, err := ioutil.ReadFile(tlsCaFile)
+	insecureSkipVerifyStr, ok := conf["tls_skip_verify"]
+	if ok {
+		insecureSkipVerify, err := strconv.ParseBool(insecureSkipVerifyStr)
```
Comment (jefferai) [inline physical/consul/consul.go]: I don't have an issue with you updating how this works, but please use parseutil.ParseBool instead.

File: physical/consul/consul.go [hunk 4]
```diff
@@ -240,65 +238,80 @@ func NewConsulBackend(conf map[string]string, logger log.Logger) (physical.Backe
...
+	if !ok {
+		tlsMinVersionStr = "tls12"
+	}
+
+	tlsMinVersion, ok := tlsutil.TLSLookup[tlsMinVersionStr]
+	if !ok {
+		return nil, fmt.Errorf("invalid 'tls_min_version'")
+	}
+
+	tlsClientConfig.MinVersion = tlsMinVersion
```
Comment (jefferai) [inline physical/consul/consul.go]: You did a whole lot of reordering which is making the diff hard to follow. Can you please change the ordering of parsing the various values back to how they were so that it's easier to make sure nothing is being left out?

File: physical/consul/consul.go (hunk 1 above)
Comment (simon-wenmouth) [reply physical/consul/consul.go]: `SetupTLSConfig` extracts the host portion from the `TLSConfig.Address` using `net.SplitHostPort` which is why I did not repeat the effort here.

```go
	if tlsConfig.Address != "" {
		server := tlsConfig.Address
		hasPort := strings.LastIndex(server, ":") > strings.LastIndex(server, "]")
		if hasPort {
			var err error
			server, _, err = net.SplitHostPort(server)
			if err != nil {
				return nil, err
			}
		}
		tlsClientConfig.ServerName = server
	}
```

File: physical/consul/consul.go (hunk 3 above)
Comment (simon-wenmouth) [reply physical/consul/consul.go]: okay, i'll rework the change to minimize its `diff`.

File: physical/consul/consul.go (hunk 4 above)
Comment (simon-wenmouth) [reply physical/consul/consul.go]: okay, i'll rework the PR to minimize the `diff`

File: website/source/docs/configuration/storage/consul.html.md (hunk 2 above)
Comment (simon-wenmouth) [reply website/source/docs/configuration/storage/consul.html.md]: This is not a duplicate - i was trying to remark on the fact that`tls_ca_path` is used  when `tls_ca_file` is not supplied (see `rootcerts.LoadCACerts`). Any suggestions for improved wording?

File: physical/consul/consul.go (hunk 3 above)
Comment (simon-wenmouth) [reply physical/consul/consul.go]: having re-read the documentation i'm going to revert to the original implementation

**Comment by jefferai** [comment]
Closing as OP seems gone.
